server/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -439,6 +439,27 @@ public ShardRouting initializeShard(
         return initializedShard;
     }
 
+    public ShardRouting relocateOrReinitializeShard(
+        ShardRouting startedShard,
+        String nodeId,
+        long expectedShardSize,
+        RoutingChangesObserver changes
+    ) {
+        if (startedShard.isSearchable()) {
+            return relocateShard(startedShard, nodeId, expectedShardSize, changes).v2();
+        } else {
+            remove(startedShard);
+            final var unassignedInfo = new UnassignedInfo(UnassignedInfo.Reason.REINITIALIZED, "relocating unsearchable shard");
+            for (ShardRouting unpromotableReplica : List.copyOf(assignedShards(startedShard.shardId()))) {
```

Comment:
... or possibly we just fail and re-initialize an unsearchable replica, but without blowing away all the search shards too.

Reply:
I think we should try to only unassign -> initialize the index/primary. If we touch replica/search shards this would not be reflected in the BalancedShard allocator balancer model